### ISSUE TITLE
Error messages do not include context to locate problematic parts in schema

### ISSUE THREAD

**Issue by FlorianLudwig** [jfinkhaeuser/prance#69]
## Expected Behaviour
Providing some context which part of my 113k line file failed.

## Actual Behaviour

```
prance validate v1.16.4.json 
Processing "v1.16.4.json"...
 -> Resolving external references.
Traceback (most recent call last):
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/bin/prance", line 8, in <module>
    sys.exit(cli())
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/cli.py", line 108, in command_invoke
    original_invoke(ctx)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/cli.py", line 197, in validate
    __validate(parser, name)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/cli.py", line 59, in __validate
    parser.parse()
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/__init__.py", line 138, in parse
    self._validate()
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/__init__.py", line 294, in _validate
    resolver.resolve_references()
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 96, in resolve_references
    self.specs = self._resolve_partial(self.parsed_url, self.specs, ())
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 207, in _resolve_partial
    changes = dict(tuple(self._dereferencing_iterator(base_url, partial, (),
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 133, in _dereferencing_iterator
    ref_value = self._dereference(ref_url, obj_path, next_recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 190, in _dereference
    value = self._resolve_partial(ref_url, value, recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 207, in _resolve_partial
    changes = dict(tuple(self._dereferencing_iterator(base_url, partial, (),
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 133, in _dereferencing_iterator
    ref_value = self._dereference(ref_url, obj_path, next_recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 190, in _dereference
    value = self._resolve_partial(ref_url, value, recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 207, in _resolve_partial
    changes = dict(tuple(self._dereferencing_iterator(base_url, partial, (),
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 133, in _dereferencing_iterator
    ref_value = self._dereference(ref_url, obj_path, next_recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 190, in _dereference
    value = self._resolve_partial(ref_url, value, recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 207, in _resolve_partial
    changes = dict(tuple(self._dereferencing_iterator(base_url, partial, (),
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 133, in _dereferencing_iterator
    ref_value = self._dereference(ref_url, obj_path, next_recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 190, in _dereference
    value = self._resolve_partial(ref_url, value, recursions)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 207, in _resolve_partial
    changes = dict(tuple(self._dereferencing_iterator(base_url, partial, (),
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/resolver.py", line 112, in _dereferencing_iterator
    ref_url, obj_path = _url.split_url_reference(base_url, refstring)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/url.py", line 120, in split_url_reference
    parsed_url = absurl(reference, base_url)
  File "/home/f.ludwig/.local/share/virtualenvs/pdk8s-oowwi4Ru/lib/python3.8/site-packages/prance/util/url.py", line 56, in absurl
    parsed = parse.urlparse(url)
  File "/usr/lib64/python3.8/urllib/parse.py", line 372, in urlparse
    url, scheme, _coerce_result = _coerce_args(url, scheme)
  File "/usr/lib64/python3.8/urllib/parse.py", line 124, in _coerce_args
    return _decode_args(args) + (_encode_result,)
  File "/usr/lib64/python3.8/urllib/parse.py", line 108, in _decode_args
    return tuple(x.decode(encoding, errors) if x else '' for x in args)
  File "/usr/lib64/python3.8/urllib/parse.py", line 108, in <genexpr>
    return tuple(x.decode(encoding, errors) if x else '' for x in args)
AttributeError: 'dict' object has no attribute 'decode'
```

## Steps to Reproduce

```
wget https://raw.githubusercontent.com/kubernetes/kubernetes/v1.16.4/api/openapi-spec/swagger.json
prance validate swagger.json
```

## Environment
- OS: Linux
- Python version: 3.8
- Swagger/OpenAPI version used: 2.0
- Backend: best

@jfinkhaeuser

**Comment by jfinkhaeuser** [comment]
It's very hard to provide context on *every error* not raised in your own code. Sorry about that.

That said, in this case I can easily provide some context. Looks like some URL you're using to refer to some object isn't parseable by urllib - which may mean it's invalid.

That is simple enough to catch and provide context on. Thanks for reporting this!

**Comment by jfinkhaeuser** [comment]
Fixed in 0.18.3

**Comment by FlorianLudwig** [comment]
Hi,

the error message now is:
```
ERROR in "v1.16.4.json" [ResolutionError]: 'dict' object has no attribute 'decode' -- Unable to parse url: {'type': 'string'}
```
which is an improvement for sure but maybe even more context would be possible? Maybe the name of the definition?

In my case `"type": "string"` is 5139 times in the json.

**Comment by jfinkhaeuser** [comment]
Oh for sure. It's just not very easy to get at at the point that the error is raised. I've been thinking about how to improve that.

In the meantime, your stack trace suggests that it's a `$ref` key that is mistakenly a dict. Since that's always wrong, even if you have 5k instances of it, you've got something to look for. Hopefully it only appears once!

**Comment by FlorianLudwig** [comment]
Thanks for the hint, did help to find it!

**Comment by FlorianLudwig** [comment]
Actually, I believe the schema is correct.

It defines a property called `$ref` which is a string. Basically defining JSONSchema in JSONSchema:

```
    "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1beta1.JSONSchemaProps": {
      "description": "JSONSchemaProps is a JSON-Schema following Specification Draft 4 (http://json-schema.org/).",
      "properties": {
        "$ref": {
          "type": "string"
        },
```

Reading the specification I am not actually sure if this is correct or not.

**Comment by jfinkhaeuser** [comment]
Well, consider `$ref` ​a reserved key if you have JSON references. I don't think that's possible. There may be a way to e.g. escape the `$` sign to make it legal.

**Comment by FlorianLudwig** [comment]
> A common misconception is that $ref is allowed anywhere in an OpenAPI specification file. Actually $ref is only allowed in places where the OpenAPI 3.0 Specification explicitly states that the value may be a reference.

-- https://swagger.io/docs/specification/using-ref/

So I guess it is not that simple that all `$ref` are references.

**Comment by jfinkhaeuser** [comment]
There's that. But OpenAPI is a conflicting spec. They also claim to be using JSON schema and JSON references, but violate those specs in places.

So OpenAPI implementations vary (hence the multiple backends, even if some of them are barely active). Most are based on JSON schema validators, because that's the easy way to implement things. Technically, those don't conform to OpenAPI precisely. You only notice it in these instances, though.

The TL;DR is, with conflicting specs, it's IMHO best to stick to the (very large) subset of things that are unambiguous. Treating `$ref` as references is kind of that, because it follows well-established specs over the somewhat more slapdash OpenAPI ones. But it's a decision, granted.

**Comment by jfinkhaeuser** [comment]
https://github.com/jfinkhaeuser/prance#a-note-on-json-references for some context.

**Comment by jfinkhaeuser** [comment]
And in the README, also maybe see the less relevant note on strict mode.

**Comment by jfinkhaeuser** [comment]
And I figured I'd update the list and move it out of the README so the README doesn't become so cluttered. https://github.com/jfinkhaeuser/prance/blob/master/COMPATIBILITY.rst

**Comment by FlorianLudwig** [comment]
Thank you for the clarifications! I probably have to read more of the specs to understand those edge cases.
